### PR TITLE
[reed_solomon] Improve Reed Solomon erasure encoding wrapper interface

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This PR makes the following changes to reed solomon encoding wrapper
- [add] `self.encode<T: BorshSerialize>` function for encoding serializable objects to parts
- [add] `self.decode<T: BorshDeserialize>` function for converting parts into object
- [remove] `self.reconstruct` function as it was too low level and difficult to understand
- [remove] `self.data_shard_count` and `self.total_shard_count`. Instead, we can rely on `epoch_manager.num_data_parts()` and `epoch_manager.num_total_parts()`

This would be used in the future to encode and decode state witness parts

### CODE REVIEW

File: core/primitives/src/reed_solomon.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        // b'aaabbbcccd' -> [Some(b'aaa'), Some(b'bbb'), Some(b'ccc'), Some(b'd00'), None, None]
+        bytes.resize(data_parts * part_length, 0);
+        let mut parts = bytes
+            .chunks_exact(part_length)
+            .map(|chunk| Some(chunk.to_vec().into_boxed_slice()))
+            .chain(itertools::repeat_n(None, self.rs.parity_shard_count()))
+            .collect_vec();
+
+        // Fine to unwrap here as we just constructed the parts
+        self.reconstruct(&mut parts).unwrap();
```
Comment (saketh-are) [inline core/primitives/src/reed_solomon.rs:47]: Previously if reconstruction failed we were printing a warn and returning. If we abort instead will that be handled appropriately up the call chain? Or, did you realize that this always succeeds and the warn on err is unnecessary?

**Review by saketh-are — COMMENTED** [review]
This looks pretty good, the main question I have is does `fn encode` always succeed? It seems like there are a couple of cases in which previously we'd print a warn and exit that are now eliminated.

File: core/primitives/src/reed_solomon.rs (hunk 1 above)
Comment (shreyan-gupta) [reply core/primitives/src/reed_solomon.rs:47]: So the unwrap here is totally fine as we just constructed the parts above and we basically guarantee they are properly constructed and in place. The reconstruct here is basically just to populate the parity shards (all the `None`s)

Comment (shreyan-gupta) [reply core/primitives/src/reed_solomon.rs:47]: The same reconstruct function is called in the decode function where we do need to handle and expose the errors.

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -1049,41 +1034,38 @@ impl ShardsManager {
...
-                        "Invalid: Wrong merkle root");
-                    ChunkStatus::Invalid
-                }
-            } else {
-                debug!(target: "chunks", "Invalid: Failed to reconstruct");
-                ChunkStatus::Invalid
-            }
-        } else {
+
+        let data_parts = self.epoch_manager.num_data_parts();
```
[the remark this answers is not in the document]
Comment (shreyan-gupta) [reply chain/chunks/src/lib.rs:1046]: I agree, but this is the convention used by reed solomon, `data_parts`, `total_parts`, `parity_parts`.

This is used not just to check reconstruction but also to determine how many parts to split the chunk into, so `num_parts_needed` would not be the most appropriate name for the second case.

File: core/primitives/src/reed_solomon.rs (hunk 1 above)
Comment (saketh-are) [reply core/primitives/src/reed_solomon.rs:47]: I see, so encode followed by decode always succeeds?

**Review by saketh-are — APPROVED** [review]

File: core/primitives/src/reed_solomon.rs (hunk 1 above)
Comment (shreyan-gupta) [reply core/primitives/src/reed_solomon.rs:47]: encode does the following: take the whole object, serialize it, split it into parts, populate the parity parts (here we call self.reconstruct) and then return the parts.

decode meanwhile does the following: take a subset of parts, try to reconstruct the data (this also calls self.reconstruct but can fail) and then deserialize and return object (deserialization can also fail)

Comment (shreyan-gupta) [reply core/primitives/src/reed_solomon.rs:47]: But yes, encode followed by decode always succeeds as long as we have "num_data_parts" number of parts.

**Pull request merged**